src/Passes/docs/base-profile-transformations.md

```diff
@@ -0,0 +1,233 @@
+  ret void
+}
+```
+
+We note that we successfully have eliminated loops, arithmetic operations, dynamic qubit allocation and alias counting - all operations which are not supported by the base profile.
+
+## Goal
+
+We envision the tool to work as a standalone commandline tool which can either validate or generate a QIR in accordance with a given profile. To validate, one would run:
+
+```language
+qat -p profile.yaml --validate unvalidated-qir.ll
+```
+
+In a similar fashion, generation is performed by adding `--generate` to the commandline:
+
+```language
+qat -p profile.yaml --generate qir.ll > qir-profile.ll
+```
+
+Default behaviour of the tool is that it always validates the generated profile. This behaviour can be disabled by
+
+```language
+qat -p profile.yaml --generate --no-validate qir.ll > qir-profile.ll
+```
+
+# Profile Specification
+
+Every profile is specified through a YAML file which defines an object at the top-level. This object must contain the fields `name` and `displayName`:
+
+```yaml
+name: profile-name
+displayName: Profile Name
+# ...
+```
+
+Additionally, top level also contains the fields `version` and `mode`. The version refers to the QIR version which forms the basis for the specification and `mode` explains how the profile is defined.
+The final two top level fields are lists named `specification` and `generation`. These contains the specification and generation producedure, respectively.
+
+## Specification
+
+The default `mode` of specification is by `feature` which means that specification describes the feature set available. Alternatively, one can specify a profile by `limitation`. As an example profile for
+
+```yaml
+name: profile-name
+displayName: Profile Name
+version: 1.0
+mode: feature
+specification:
+  functions:
+    - __quantum__qis__toffoli__body
+    - __quantum__qis__cnot__body
+    - __quantum__qis__cz__body
+    - __quantum__qis__h__body
+    - __quantum__qis__mz__body
+    - __quantum__qis__reset__body
+    - __quantum__qis__rx__body
+    - __quantum__qis__ry__body
+    - __quantum__qis__rz__body
+    - __quantum__qis__s__body
+    - __quantum__qis__s_adj
+    - __quantum__qis__t__body
+    - __quantum__qis__t__adj
+    - __quantum__qis__x__body
+    - __quantum__qis__y__body
+    - __quantum__qis__z__body
+  instructions:
+    - call
+    - br
+    - ret
+    - inttoptr
+# ...
+```
+
+This specification describes that 16 quantum instructions are available and 4 classical operations of the full QIR spec. Contrary, a specification by `limitation` could be as follows:
+
+```yaml
+name: profile-name
+displayName: Profile Name
+version: 1.0
+mode: limitation
+specification:
+  functions:
+    - __quantum__rt__array_update_alias_count
+    - __quantum__rt__array_update_reference_count
+  instructions:
+    - br
+# ...
+```
+
+This profile specifies a system that does not allow reference and alias counting and neither have support for branching, but otherwise has the full QIR vesion 1.0 available.
```

Comment:
I am not sure that any combination of limitations actually will lead to a well-defined profile - I believe there are essentially dependencies/implications between certain limitations. I can also imaging that we have certain features be exclusive (you can have either feature A or feature B but never both at the same time). What is your thinking here? Are these features and limitations in any way validated?

Reply:
No, I realised this after discovering that QIR does not specify all the functions, so this will need to be rethought. It can still be helpful to have spec by exclusion (even if i don't really like this method of defining what something is. But for instance, you could "include all integer operations, but exclude shift operations" rather than specifying all operations. 

We could have mutually exclusive features. I have not given it much thought, so I will think a bit about it. Do you have a good example?